packages/marshal/src/helpers/promise.js

```diff
@@ -0,0 +1,45 @@
+// @ts-check
+
+/// <reference types="ses"/>
+
+import { isPromise } from '@endo/promise-kit';
+import { assertChecker, PASS_STYLE } from './passStyle-helpers.js';
+import { assertSafePromise } from './safe-promise.js';
+
+const { details: X, quote: q } = assert;
+const { ownKeys } = Reflect;
+const { getPrototypeOf, prototype: objectPrototype } = Object;
+
+/**
+ *
+ * @type {import('./internal-types.js').PassStyleHelper}
+ */
+export const PromiseHelper = harden({
+  styleName: 'promise',
+
+  canBeValid: (candidate, check) => {
+    return (
+      (isPromise(candidate) ||
+      candidate[PASS_STYLE] === 'promise' ||
+      check(false, X`Missing ${q(PASS_STYLE)}: ${candidate}`))
```

Comment:
The `assertValid` below is not close to adequate input validation. For example, `candidate[PASS_STYLE]` might be an accessor that changes its value each time it is read. However, if we replace these two lines in `canBeValid` with

```suggestion
      checkTagRecord(candidate, 'promise', check)
```

then we get a more accurate check with better diagnostics.

Reply:
I have adopted this in `assertValid`, although it does make construction rather cumbersome:
```js
Object.create(Object.prototype, {
  [PASS_STYLE]: { value: 'promise' },
  { [Symbol.toStringTag]: value: 'Pseudo-promise' },
});
```